### ISSUE TITLE
SavedModel exporting fails on RNNs by setting wrong input dtypes

### ISSUE THREAD

**Issue by leandro-gracia-gil** [tensorflow/tensorflow#41815]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: 3.7.7
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
When trying to export concrete functions in a Keras model to the SavedModel format, if a function uses `tf.keras.layers.RNN`, the dtypes of the RNN cell input arguments are wrong. In particular, actual dtypes seem to be ignored and replaced with tf.float32. This does not happen when calling the same concrete function manually, but only when trying to export it.

**Describe the expected behavior**
The input dtypes should be preserved, allowing to export the model. At the very least the behavior should be consistent with calling the concrete function.

**Standalone code to reproduce the issue**
```python
import tensorflow as tf

class TestRNNCell(tf.keras.layers.Layer):
  def __init__(self):
    super().__init__()
    self.units = 10
    self.state_size = 20

  def call(self, indices, state):
    # This assertion fails.
    assert indices.dtype == tf.int32
    # If the assertion is removed, this line fails with:
    # TypeError: Value passed to parameter 'indices' has DataType float32 not in list of allowed values: int32, int64
    output = tf.gather(tf.range(5), indices)
    return output, state

class TestModel(tf.keras.Model):
  def __init__(self):
    super().__init__()
    self.rnn = tf.keras.layers.RNN(TestRNNCell())

  @tf.function
  def do_stuff(self, indices):
    assert indices.dtype == tf.int32
    return self.rnn(indices)

model = TestModel()
tf.saved_model.save(model, 'test_model', signatures={
  'do_stuff': model.do_stuff.get_concrete_function(
      indices=tf.TensorSpec([None, None, 5], tf.int32))
})
```

Using `model.save` instead of `tf.saved_model.save` does not help either. Here's a version that fails just in the same way.
```python
class TestModel(tf.keras.Model):
  # [Continuing the class above]

  def call(self, indices):
    assert indices.dtype == tf.int32
    return self.rnn(indices)

model = TestModel()

# This works fine.
model(tf.zeros([10, 10, 5], tf.int32))

# This fails because the RNN inputs are now tf.float32.
model.save('test_model', save_format='tf')
```

Any possible workaround for this would be much appreciated. I cannot figure out any since this problem only affects SavedModel exporting and not the actual concrete functions being exported.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/2f6efc1ad81e0c6f239d0f9f0fd29805/41815-2-3.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/c2fd2680adb7cd75de9e8e455b1b4988/41815-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by k-w-w** [comment]
The error is a bit hard to follow, but you won't get the error if you set the dtype of the RNN layer:

```
    self.rnn = tf.keras.layers.RNN(TestRNNCell(),dtype=tf.int32)
```
